plugins/ports-plugin/src/ports-plugin.ts

```diff
@@ -10,191 +10,326 @@
-    if (result && result.title === 'yes') {
-        // takes first available port
-        const workspacePort = redirectPorts.pop()!;
-
-        // start a new listener
-        const portRedirectListener = new PortRedirectListener(parseInt(workspacePort.portNumber, 10), 'localhost', port.portNumber);
-        portRedirectListener.start();
-
-        // store port taken
-        const busyPort = { portRedirectListener, workspacePort };
-        redirectListeners.set(port.portNumber, busyPort);
-
-        // show redirect
-        const redirectInteractions: MessageItem[] = [{ title: 'Open Link' }];
-        const msg = `Redirect is now enabled on port ${port.portNumber}. External URL is ${workspacePort.url}`;
-        const resultShow = await theia.window.showInformationMessage(msg, ...redirectInteractions);
-        if (resultShow && resultShow.title === 'Open Link') {
-            theia.commands.executeCommand('mini-browser.openUrl', workspacePort.url);
+    /**
+     * Prompt user to create a port redirect for the specific port
+     * @param port the port that needs to be redirected
+     * @param redirectMessage the message if there are 'free ports' in workspace
+     * @param errorMessage  if no free port are available
+     */
+    async askRedirect(port: ListeningPort, redirectMessage: string, errorMessage: string): Promise<void> {
+
+        // grab a free redirect
+        if (this.redirectPorts.length === 0) {
+            await theia.window.showErrorMessage(errorMessage, { modal: true });
+            return;
         }
-    }
-}
-
-// Callback when a new port is being opened in workspace
-async function onOpenPort(port: Port): Promise<void> {
 
-    // skip excluded
-    if (excludedPorts.includes(port.portNumber)) {
-        // this port is excluded so just print a notice but does not propose a redirect
-        outputChannel.appendLine(`Ignoring excluded port ${port.portNumber}`);
-        return;
+        const interactions: MessageItem[] = [{ title: 'yes' }, { title: 'no' }];
+        const result = await theia.window.showInformationMessage(redirectMessage, ...interactions);
+        if (result && result.title === 'yes') {
+            // takes first available port
+            const endpoint = this.redirectPorts.pop()!;
+
+            // start a new listener
+            const portRedirectListener = new PortRedirectListener(endpoint.targetPort, 'localhost', port.portNumber);
+            portRedirectListener.start();
+
+            // store port taken
+            const forwardedPort = { portRedirectListener, endpoint };
+            this.redirectListeners.set(port.portNumber, forwardedPort);
+            this.updateEndpoints();
+
+            // show redirect
+            const redirectInteractions: MessageItem[] = [{ title: 'Open In New Tab' }];
+            if (endpoint.protocol === 'https') {
+                redirectInteractions.push({ title: 'Open Link' });
```

Comment:
this part will wait a smallest specific PR as it was breaking all happy path tests. so decided to make the change with happy path enhancement

Reply:
Sorry, I don't understand your comment: is this also because it would break the happy path test? If so, the same remark about a todo and an issue against the happy path test would apply.

Reply:
I still don't get it 🤷